prime-router/docs/design/design/auth/auth-design.md

```diff
@@ -0,0 +1,431 @@
+            "endpoint": "/api/waters",
+            "methods": {
+                "POST": {
+                    "scopes": ["report"]
+                }
+            }
+        },
+        {
+            "endpoint": "api/history/{orgName}/submissions",
+            "methods": {
+                "GET": {
+                    "scopes": ["user"]
+                }
+            }
+        }
+    ]
+}
+```
+
+![Reverse Proxy Microservice](reverseProxyMicroservice.png)
+
+### Use libraries with limited code in shared project
+
+This approach would attempt to use existing libraries to annotate request mappings in our Spring app.
+
+We would have to write custom code in the shared project to handle configuration and authn/authz for ResportStream
+endpoints that still live in our Azure Functions app.
+
+Pros
+- Simple implementation for a small project like ours
+- Authn/authz annotations would be directly adjacent to the endpoint mappings they protect
+- Can handle complex authorization requirements for specific endpoints with more ease
+- Simplifies project structure without new deployments
+- No additional devops support
+
+Cons
+- Less scalable solution
+- Would have to deal with handling authn/authz with 2 different web frameworks
+  -  Easy for Spring but could require some custom work for Azure Functions
+
+A typical report request would look like this
+- Client makes request to authorization server (or us in the ReportStream implementation solution) to retrieve token with their own credentials
+- Client uses retrieved token with request to ResportStream
+- Request hits ReportStream resource directly
+- Depending on our implementation, we decide if that request is legitimate and allowed to access the designated endpoint
+    - In cloud-based solutions, we would make calls out to an external authz server
+      - We could also introspect the token locally if speed is a factor but its more code to write/maintain
+    - In a ReportStream implementation solution, we would introspect certain cookies and headers and compare them to what we have in a datastore
+- If the request is unauthenticated or unauthorized, short-circuit the request to a 401
+- If the request is authenticated and authorized for the resource, continue to the business logic
+
+Code example:
+```kotlin
+// Spring project
+// has read scope
+@GetMapping("/api/v1/protected/endpoint")
+@PreAuthorize("hasAuthority('SCOPE_read')") // scope 
+fun protectedReadEndpoint() {
+    // read that anyone with the read scope can see
+}
+
+// Spring project
+// has admin authority
+@PostMapping("/api/v1/protected/endpoint")
+@PreAuthorize("hasAuthority('Admin')") // role
+fun protectedWriteEndpoint() {
+    // write that only admins can do
+}
+
+// Azure functions project
+// for an Azure HTTP endpoint we could write a function to handle the authn/authz check
+fun reportsEndpoint() {
+    checkAuthn(request) { accessToken: JWT ->
+        requiredAuthzScope(accessToken, listOf("requiredScope")) {
```

Comment:
Comment applies everywhere but putting it here:

We should remember that scope isn't the only criteria for determining authorization, in ReportStream. In the case of submitting reports, the client should have a `SCOPE` like "write" but ALSO their `CLAIMS` must include the `organization.sender` they are trying to submit as. Same applies for history endpoint: just because a client has the proper "read" SCOPE, doesn't mean they can get the history info for a report they didn't submit.

Reply:
Okay never mind you discuss this in the section titled `### Scopes and claims` below but I will leave this comment here for documentation purposes. No action required.